packages/react/src/extend-theme.ts

```diff
@@ -21,45 +25,118 @@ type DeepThemeExtension<BaseTheme, ThemeType> = {
     : CloneKey<ThemeType, Key>
 }
 
-export type ThemeOverride = Partial<ChakraTheme> &
-  DeepThemeExtension<Theme, ChakraTheme>
+export type ThemeOverride<BaseTheme = Theme> = Partial<ChakraTheme> &
+  DeepThemeExtension<BaseTheme, ChakraTheme> &
+  Dict
+
+export type ThemeExtension<Override extends ThemeOverride = ThemeOverride> = (
+  themeOverride: Override,
+) => Override
+
+export type BaseThemeWithExtensions<
+  BaseTheme extends ChakraTheme,
+  Extensions extends readonly [...any]
+> = Extensions extends [infer L, ...infer R]
+  ? L extends (...args: any[]) => any
+    ? ReturnType<L> & BaseThemeWithExtensions<BaseTheme, R>
+    : L & BaseThemeWithExtensions<BaseTheme, R>
+  : BaseTheme & Extensions
 
 /**
- * Function to override or customize the Chakra UI theme conveniently
- * @param overrides - Your custom theme object overrides
- * @param baseTheme - theme to customize
+ * Function to override or customize the Chakra UI theme conveniently.
+ * First extension overrides the baseTheme and following extensions override the preceding extensions.
+ *
+ * @example with only overrides:
+ * import { extendTheme } from '@chakra-ui/react'
+ *
+ * const customTheme = extendTheme({
+ *   colors: {
+ *     brand: {
+ *       500: "#b4d455",
+ *     },
+ *   },
+ * })
+ *
+ *
+ * @example with extension:
+ * import { theme, extendTheme, withDefaultColorScheme } from '@chakra-ui/react'
+ *
+ * const customTheme = extendTheme(
+ *   {
+ *     colors: {
+ *       brand: {
+ *         500: "#b4d455",
+ *       },
+ *     },
+ *   },
+ *   withDefaultColorScheme({ colorScheme: "red" }),
+ *   theme,
+ * )
  */
 export function extendTheme<
   BaseTheme extends ChakraTheme = Theme,
-  Overrides extends ThemeOverride = ThemeOverride
+  Extensions extends (
+    | BaseTheme
+    | ThemeOverride<BaseTheme>
+    | ThemeExtension<ThemeOverride<BaseTheme>>
+  )[] = (ThemeOverride<BaseTheme> | ThemeExtension<ThemeOverride<BaseTheme>>)[]
 >(
-  overrides: Overrides,
-  baseTheme: BaseTheme = (defaultTheme as unknown) as BaseTheme,
-): BaseTheme & Overrides {
-  function customizer(
-    source: unknown,
-    override: unknown,
-    key: string,
-    object: any,
+  ...extensions: [...Extensions]
+): BaseThemeWithExtensions<BaseTheme, Extensions> {
+  const overrides = [...extensions]
+    .slice(0, extensions.length - (extensions.length > 1 ? 1 : 0))
+    .reverse()
+
+  const baseTheme =
+    extensions.length > 1 ? extensions[extensions.length - 1] : defaultTheme
+
+  if (!isChakraTheme(baseTheme) || !overrides.length) {
+    warn(
+      `Last parameter needs to be a valid \`ChakraTheme\`. Please check your extendTheme function call from @chakra-ui/react.`,
+    )
+  }
```

Comment:
It seems like we could improve the DX and reduce the complexity of implementation if we made the same assumption about `baseTheme` as we do with `extendTheme` currently: the last argument can be a `theme` to modify, but if it's not, we use the default Chakra theme.

```jsx
const theme = extendTheme({
  // theme stuff
})

const theme = extendTheme(
  compose(),
  some(),
  things(),
  withNo(),
  theme()
)
```

This is similar to how [`customize-cra`](https://github.com/arackaf/customize-cra) works, where you're just defining the pipeline of functions that will be executed on the base theme, at least by default.

Reply:
Seems like from looking at the related docs, we have to make the last object a theme because the last item in the list could theoretically be a theme customization object that passes as a theme, which would obviously break things. 

That feels awkward to me, and I'm wondering if we could make this significantly better by making it a totally new API rather than modifying the existing one, so we don't have to maintain any backwards compatibility.

I'm have lots of thoughts here but I'm not sure this is the right place to share them. Maybe we could hang out and chat about it @TimKolberger?

Reply:
Moving this logic to a new function would ease the pain of backwards compat! There are pros and cons about this approach - let's chat about this :)

Reply:
The `baseTheme` is now optional for all use cases (with 1 or more overrides/extensions).
I updated the implementation of `extendTheme` to be more readable. With this in place I am not quite sure if another function would increase the maintainability. Let me know what you think :)